orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,402 @@
+    const TensorShape shape{
+        tensor_proto->dims().data(), static_cast<size_t>(tensor_proto->dims().size())};
+
+    std::vector<char> tensor_buffer{};
+    tensor_buffer.resize(element_size * shape.Size());
+
+    const MemBuffer mem_buffer{tensor_buffer.data(), tensor_buffer.size(), cpu_alloc_info};
+
+    OrtValue ort_value;
+
+    ORT_RETURN_IF_ERROR(onnxruntime::utils::TensorProtoToMLValue(
+        Env::Default(), nullptr, *tensor_proto, mem_buffer,
+        ort_value));
+
+    name_to_ort_value.emplace(tensor_proto->name(), ort_value);
+    tensor_buffers.emplace_back(std::move(tensor_buffer));
+  }
+
+  return Status::OK();
+}
+
+const std::vector<std::string> ParseStringData(
+    const ONNX_NAMESPACE::TensorProto* tensor_proto) {
+  ORT_ENFORCE(!tensor_proto->has_data_type() ||
+                  tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED,
+              "Invalid string data type.");
+  ORT_ENFORCE(tensor_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING,
+              "ParseStringData type mismatch for tensor ");
+
+  ORT_ENFORCE(!(tensor_proto->has_data_location() &&
+                tensor_proto->data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL),
+              "Cannot parse string data from external tensors.");
+
+  ORT_ENFORCE(!tensor_proto->has_raw_data(),
+              "stringcontent is required to be stored in repeated"
+              "bytes string_data field. raw_data type cannot be string.");
+
+  std::vector<std::string> res;
+  const auto& data = tensor_proto->string_data();
+  int expected_size = 1;
+  for (int i = 0; i < tensor_proto->dims_size(); ++i) {
+    expected_size *= tensor_proto->dims(i);
+  }
+
+  ORT_ENFORCE(tensor_proto->dims_size() != 0 && data.size() != expected_size, "Data size mismatch.");
+  res.insert(res.end(), data.begin(), data.end());
+  return res;
+}
+
+Status CheckpointUtils::OrtSaveInternal(
+    const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+    const PathString& checkpoint_path) {
+  std::unordered_map<std::string, OrtValue> name_to_ort_value;
+  ORT_RETURN_IF_ERROR(CreateOrtValuesFromTensorProtos(tensor_protos, name_to_ort_value));
+
+  auto cpu_data_transfer = std::make_unique<CPUDataTransfer>();
+  DataTransferManager data_transfer_mgr;
+  auto st = data_transfer_mgr.RegisterDataTransfer(std::move(cpu_data_transfer));
+  if (!st.IsOK()) {
+    return st;
+  }
+
+  CheckpointStates states;
+  states.module_checkpoint_states.train_session_data_transfer_mgr_ = &data_transfer_mgr;
+  auto& named_parameters = states.module_checkpoint_states.named_parameters;
+  for (auto it = name_to_ort_value.begin(); it != name_to_ort_value.end(); ++it) {
+    named_parameters.insert({it->first, std::make_shared<Parameter>(it->first, it->second)});
+  }
+
+  ORT_RETURN_IF_ERROR(OrtSaveInternal(states, checkpoint_path));
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtSaveInternal(
```

Comment:
Since the user provides a single object `states`, is it possible for us to put the contents of that entire object into a single file (instead of putting them inside different files under the same folder)?
Alternatively, if we think that these should reside in different files, do you think we could let the user decide where these different files should be stored? For example, the user could do:
```cpp
auto module_state_dict = module.GetStateDict();
ORTSave(module_state_dict, "module_state_dict.ckpt");
auto optimizer_state_dict = optimizer.GetStateDict();
ORTSave(optimizer_state_dict, "optimizer_state_dict.ckpt");
...
```
Maybe this is not important at all, just wanted to know your thoughts.

Reply:
> Since the user provides a single object states, is it possible for us to put the contents of that entire object into a single file (instead of putting them inside different files under the same folder)?

Technically, it is do-able, but it would be less clear of the data structures we saved, once we mixed all trainable weights/untrainable weights/optimizer states (momentum_1, and others) together, and use offset and size to record what this part is, what that part is. For most of scenarios, I believe optimizer states will not be needed, and splitting them is a good fit for evaluation/inference ckpt requirements. 

At this stage, per the discussion with @askhade, here is our thinking:
1. Optimizer state most of time would not be saved by users. 
2. Parameter state will be stored into 1 ~ 2 protobuf files by default, unless the file potentially > 2GB (hit protobuf 2GB size limit). 

An alternative to Protobuf, there is ORTFormat, I actually would prefer to use that if we aim to save in a single file for all ckpt states.

Reply:
ok, this makes sense.